experimental/oak_async/src/grpc.rs

```diff
@@ -0,0 +1,131 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! gRPC related types for asynchronous processing.
+
+use log::error;
+use oak::{
+    grpc::{ChannelResponseWriter, Code, WriteMode},
+    OakError,
+};
+
+/// Writes exactly one message on a gRPC response channel, setting an `UNKNOWN` error if unused.
+///
+/// This type is used for gRPC methods where the return type is non-streaming. Instances must be
+/// consumed using [`send`](#method.send). In exceptional circumstances the
+/// [`into_inner`](#method.into_inner) method may be used to access the underlying
+/// `ChannelResponseWriter`.
+///
+/// If neither [`send`](#method.send) nor [`into_inner`](#method.into_inner) is called, an `UNKNOWN`
+/// error will be written as the response upon [`Drop`](#impl-Drop).
+#[must_use = "exactly one message should be sent. Use into_inner for advanced use cases"]
+pub struct OneshotWriter<T> {
+    writer: Option<ChannelResponseWriter>,
+    _msg: core::marker::PhantomData<T>,
+}
+
+impl<T: ::prost::Message> OneshotWriter<T> {
+    // Only for use by the Oak Service Generator.
+    #[doc(hidden)]
+    pub fn new(writer: ChannelResponseWriter) -> OneshotWriter<T> {
+        OneshotWriter {
+            writer: Some(writer),
+            _msg: core::marker::PhantomData,
+        }
+    }
+
+    /// Send the response, automatically closing the channel.
+    pub fn send(mut self, msg: &T) -> Result<(), OakError> {
+        self.writer.take().unwrap().write(msg, WriteMode::Close)?;
+        Ok(())
+    }
+
+    /// Consumes this writer, returning a `ChannelResponseWriter` in its place.
+    pub fn into_inner(mut self) -> ChannelResponseWriter {
+        self.writer.take().unwrap()
+    }
+}
+
+impl<T> Drop for OneshotWriter<T> {
+    /// Close the channel with an `Unknown` status if no message was sent.
+    fn drop(&mut self) {
+        if let Some(writer) = self.writer.take() {
+            error!("drop() called on OneshotWriter");
+            if let Err(e) = writer.close(Err(oak::grpc::build_status(Code::Unknown, ""))) {
+                error!("Failed to close OneshotWriter in Drop: {}", e);
+            }
+        }
+    }
```

Comment:
This and the `must_use` seem a bit too opinionated, why send out a generic error if the caller did not want to send anything at all? I would leave the semantics of `OneShotWriter` closer to "at most once" rather than "exactly once", especially since in general the "exact" part can never be enforced correctly.

Reply:
There are all sort of other things that get in the way of delivering messages, most notably network connectivity, or queues / buffers being full. If the reader on the other hand of this channel expects exactly one message, there will still be cases in which it may not receive one, so the expectation is unrealistic, no matter how much effort the sender puts into it.

In short: yes, sending nothing at all should be considered a valid response IMO.

Reply:
I agree with you, the reader should not unconditionally expect a message to be returned from an RPC call. 
We are however looking at the writer's perspective here, and I am having a hard time imagining a situation where not sending any reply at all would be considered expected behaviour for a server.

Maybe this is just my lack of imagination though, do you have an example of when it would be valid for a server to return nothing?

In my opinion, making this very hard to do helps write correct programs. The 0.1% of people that want to do this anyway and have a valid reason can always use `into_inner`.

Reply:
Maybe a node has two writers available, and decides to send a message to one or the other based on some condition. I would just keep things as minimal as possible instead of adding extra checks that for the most part I don't think they are going to catch any interesting mistakes (who accidentally forgets to send a message?). Also I think at some point we may unify these with standard types from some other crate, so having a clear 1:1 correspondence in behaviour may help. Anyways, feel free to submit as it is if you prefer, your call, thanks for the work!

Reply:
The example you give would require two incoming requests to be entangled in such a way that we pick one of the two. In that case I think that even the request that we chose to drop would like to know that instead of a connection loss it was not the best candidate.

I'm inclined to keep the `must_use` annotation and see how that plays out when rewriting the examples. If it turns out that there's no need for it or it makes things more complicated than they should be, we can revisit.